backupccl: reduce allocations in spansForAllTableIndexes

stevendanna opened the pull request:
The table descriptor builder API copies the descriptor and does a
number of allocations so that it can answer any question one might
have about the table.

However, in this function, we only care about collecting the active
indexes and constructing spans for them.  Doing this manually is
straightforward and reduces the number of allocations considerably
when processing a large number of revisions.

```
BEFORE: BenchmarkSpansForAllTableIndexes/revisionCount=15000-16     16  62820672 ns/op  63661512 B/op     240618 allocs/op
AFTER:  BenchmarkSpansForAllTableIndexes/revisionCount=15000-16     49  22688142 ns/op   7289864 B/op      75030 allocs/op
```

This brings the benchmark case for a large number of revisions into
line with the case of a single table that just happens to have a large
number of indexes.

Release note: None

stevendanna committed:
backupccl: reduce allocations in spansForAllTableIndexes

The table descriptor builder API copies the descriptor and does a
number of allocations so that it can answer any question one might
have about the table.

However, in this function, we only care about collecting the active
indexes and constructing spans for them.  Doing this manually is
straightforward and reduces the number of allocations considerably
when processing a large number of revisions.

```
BEFORE: BenchmarkSpansForAllTableIndexes/revisionCount=15000-16     16  62820672 ns/op  63661512 B/op     240618 allocs/op
AFTER:  BenchmarkSpansForAllTableIndexes/revisionCount=15000-16     49  22688142 ns/op   7289864 B/op      75030 allocs/op
```

This brings the benchmark case for a large number of revisions into
line with the case of a single table that just happens to have a large
number of indexes.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/81754)
<!-- Reviewable:end -->

dt commented on pkg/ccl/backupccl/backup_planning.go:
```diff
@@ -125,21 +126,25 @@ var featureFullBackupUserSubdir = settings.RegisterBoolSetting(
 func getPublicIndexTableSpans(
-	table catalog.TableDescriptor, added map[tableAndIndex]bool, codec keys.SQLCodec,
+	table *descpb.TableDescriptor, added map[tableAndIndex]bool, codec keys.SQLCodec,
 ) ([]roachpb.Span, error) {
```
does this still need the error return?

stevendanna replied:
Nope, fixed. Thanks!

dt reviewed (approved):
looks ok to me but I'll defer to @ajwerner in case he'd rather keep these desc/span types of ops under the catalog packages (in which case we could spansForallIndexes in its entirety?)

ajwerner commented:
As a compromise I'd be cool with a function or method in descpb that iterates the public indexes like below and leave the span logic up here. Thoughts?

```go
package descpb

// ForEachPublicIndex is exported to provide low-overhead access to the set of
// public indexes in a table descriptor for us in backup planning. 
func (d *TableDescriptor) ForEachPublicIndex(f func(*IndexDescriptor)) {
    f(&d.PrimaryIndex)
    for i := range d.Indexes {
        f(&d.Indexes[i])
    }
}
```

stevendanna commented:
@ajwerner I like it.  

This function still looks like it accounts from 5-10% of the running time of the benchmark. But, from here I think it makes sense to look at a real-world test rather than this microbenchmark.

ajwerner commented:
Is it slower than when you didn't have it?

stevendanna committed:
backupccl: avoid small allocations in spansForAllTableIndexes

This small refactor removes the intermediate slices used to pass the
index spans around.

```
name                                            old time/op    new time/op    delta
SpansForAllTableIndexes/revisionCount=15000-16    21.7ms ± 6%    21.0ms ± 6%   -2.99%  (p=0.005 n=20+20)

name                                            old alloc/op   new alloc/op   delta
SpansForAllTableIndexes/revisionCount=15000-16    8.13MB ± 0%    7.41MB ± 0%   -8.85%  (p=0.000 n=19+19)

name                                            old allocs/op  new allocs/op  delta
SpansForAllTableIndexes/revisionCount=15000-16     75.6k ± 0%     60.6k ± 0%  -19.84%  (p=0.000 n=19+20)
```

Release Note: none

stevendanna committed:
backupccl: prealloc slice in spansForAllTableIndexes

Release note: None

stevendanna committed:
backupccl: use fast llrb tree insertion in spansForAllTableIndexes

Since we only use this tree inside this function and can easily audit
all uses of it, using fast insertion here seems safe and appears to be
a win:

```
name                                                    old time/op    new time/op    delta
SpansForAllTableIndexes/secondaryIndexesCount=15000-16    17.4ms ± 3%    13.6ms ± 4%  -22.24%  (p=0.000 n=17+20)
SpansForAllTableIndexes/revisionCount=15000-16            19.1ms ± 3%    17.4ms ± 8%   -8.64%  (p=0.000 n=20+19)

name                                                    old alloc/op   new alloc/op   delta
SpansForAllTableIndexes/secondaryIndexesCount=15000-16    3.96MB ± 0%    3.96MB ± 0%     ~     (p=0.445 n=20+20)
SpansForAllTableIndexes/revisionCount=15000-16            3.96MB ± 0%    3.96MB ± 0%     ~     (p=0.354 n=20+19)

name                                                    old allocs/op  new allocs/op  delta
SpansForAllTableIndexes/secondaryIndexesCount=15000-16     60.6k ± 0%     60.6k ± 0%     ~     (p=0.666 n=20+20)
SpansForAllTableIndexes/revisionCount=15000-16             60.6k ± 0%     60.6k ± 0%     ~     (p=0.266 n=20+19)
```

Release note: None

stevendanna commented:
@ajwerner Yesterday it was looking a bit slower, but I think that might have been some test environment issues.  This morning benchstat says it is fine.

@dt Apologies, I added a few commit here.

Overall, the difference from master is:


```
name                                                    old time/op    new time/op    delta
SpansForAllTableIndexes/secondaryIndexesCount=15000-16    19.7ms ± 4%    13.6ms ± 8%  -31.00%  (p=0.000 n=38+38)
SpansForAllTableIndexes/revisionCount=15000-16            56.9ms ± 2%    16.1ms ± 6%  -71.74%  (p=0.000 n=37+37)

name                                                    old alloc/op   new alloc/op   delta
SpansForAllTableIndexes/secondaryIndexesCount=15000-16    11.6MB ± 0%     4.0MB ± 0%  -65.78%  (p=0.000 n=40+40)
SpansForAllTableIndexes/revisionCount=15000-16            63.7MB ± 0%     4.0MB ± 0%  -93.77%  (p=0.000 n=39+40)

name                                                    old allocs/op  new allocs/op  delta
SpansForAllTableIndexes/secondaryIndexesCount=15000-16     60.6k ± 0%     60.6k ± 0%   -0.07%  (p=0.000 n=40+40)
SpansForAllTableIndexes/revisionCount=15000-16              241k ± 0%       61k ± 0%  -74.83%  (p=0.000 n=39+39)
```

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 3 files at r2, 1 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81754#-:-N34ePK24cZGrWSn-ZIx:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @dt)



<!-- Sent from Reviewable.io -->

stevendanna commented:
bors r=dt,ajwerner

stevendanna commented:
blathers backport 22.1
